lib/block-supports/custom-css.php

```diff
@@ -0,0 +1,72 @@
+<?php
+/**
+ * Align block support flag.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Registers the custom css block attribute for block types that support it.
+ *
+ * @param WP_Block_Type $block_type Block Type.
+ */
+function gutenberg_register_custom_css_support( $block_type ) {
+	$has_custom_css_support = false;
+	if ( property_exists( $block_type, 'supports' ) ) {
+		$has_custom_css_support = gutenberg_experimental_get( $block_type->supports, array( 'customCSS' ), false );
+	}
+	if ( $has_custom_css_support ) {
+		if ( ! $block_type->attributes ) {
+			$block_type->attributes = array();
+		}
+
+		if ( ! array_key_exists( 'css', $block_type->attributes ) ) {
+			$block_type->attributes['css'] = array(
+				'type' => 'string',
+			);
+		}
+	}
+}
+
+/**
+ * Add generated CSS class and the stylesheet.
+ *
+ * @param array         $attributes       Comprehensive list of attributes to be applied.
+ * @param array         $block_attributes Block attributes.
+ * @param WP_Block_Type $block_type       Block Type.
+ *
+ * @return array Block custom CSS classes.
+ */
+function gutenberg_apply_custom_css_support( $attributes, $block_attributes, $block_type ) {
+	$has_custom_css_support = false;
+	if ( property_exists( $block_type, 'supports' ) ) {
+		$has_custom_css_support = gutenberg_experimental_get( $block_type->supports, array( 'customCSS' ), false );
+    }
+	if ( $has_custom_css_support && isset( $block_attributes['css'] ) && ! empty(  $block_attributes[ 'css' ] ) ) {
+        $generated_class_name = uniqid( 'wp-block-css-' );
+        wp_add_inline_style( 
+            'wp-block-custom-css', 
+            str_replace( ':self', '.' . $generated_class_name, $block_attributes[ 'css' ] )
+        );
+        $attributes['css_classes'][] = $generated_class_name;
+	}
+
+	return $attributes;
+}
+
+/**
+ * This style is used to enqueue the dynamic block custom CSS styles.
+ */
+function gutenberg_register_custom_css_style() {
+    wp_register_style('wp-block-custom-css', false);
```

Comment:
I'm using an empty style to enqueue the different block custom CSS. 
Let me know if there are better ways to do that.

Reply:
Ideally we wouldn't enqueue anything... These would be a lot better as inline styles in the element itself. :+1: 
If the user adds styles to a specific block, then they want those styles applied and they should therefore override any global CSS from stylesheets.

Reply:
> Ideally we wouldn't enqueue anything... These would be a lot better as inline styles in the element itself. 👍

This assumes there's a way to alter the markup of the block and add this somewhere there. IMO it's a bit more fragile because of the DOM changes needed.

Reply:
I was thinking of handling this the same way we handle extra, user-defined classes... The logic we follow for that is pretty robust and we could do the same for custom-css

Reply:
We don't have extra markup inside the block wrapper though, also, we can't assume the wrapper accept children or is a self closing tag.

Reply:
The way I was envisioning it would be a bit more restrictive... 
So users wouldn't be able to type full CSS like `:self{color:red;}`. Instead they'd only write the rules (no selector) like `color:red;`. 
That would then be properly added to the element as inline CSS using `style=""` in the wrapper itself, there's no need to alter any other markup, children etc. It would be a lot easier to sanitize too (basically `esc_attr` on the PHP side and an equivalent in JS).
That's why I envisioned it similar to the "Additional CSS class(es)" field... The implementation would be pretty much identical.
So the field wouldn't be so much a "custom CSS" field... it would be "custom styles" - the difference being it will only contain the actual styles so no selectors etc, can't be considered full CSS since there's nothing cascading about it.

Reply:
Yes, I thought about that, the problem with that though, is that you can alter pseudo-states and things like that.

Reply:
True... But if someone is knowledgeble enough to know what pseudo-elements are, they can surely add a custom class to their element and then add whatever CSS they want... They can even add CSS in a custom HTML block. :wink: 
A feature for custom-styles should appeal and be usable by the masses. For more advanced features there are more advanced solutions to adding CSS

Reply:
It's added to the "Advanced" section for a reason. I think that's where the disagreement lies, I do think it's a feature for advanced users regardless of how it's implemented. It's something to be used to create designs that are not possible with regular controls. Imagine creative things like animations, filters, rotations... in predefined block patterns. 

You might be able to achieve these things with dedicated plugins... but for adhoc situations where you need these things, you don't want to install a plugin that supports animations, rotations for Core blocks and create potential invalidation.

Reply:
I agree with Riad that this feature is useful as it stands, even when there are other more generic tools to achieve similar results.

This feature allows styles to be tied specifically to a block, regardless of the chosen selectors. Providing `:self` is very useful, but it is even better when we allow advanced use cases — not just pseudo-states and nested elements — like (why not) selecting elements outside of :self.